### DISCUSSION TITLE
Native-sized floating-point numeric types (`nfloat`) and native-sized math library (`NMath`)

### DISCUSSION THREAD

**Discussion by tvardero** [dotnet/csharplang#6553] [Language Ideas]
Since C# 9 we have `nint` and `nuint`, but what about `nfloat`? As well as `NMath` library for math operations (sin, cos, round, etc.).
I see that Xamarin.Mac SDK [has it already](https://learn.microsoft.com/en-us/dotnet/api/system.nfloat?view=xamarin-mac-sdk-14).

Current approach is to use this code:
```c#
#if REAL_T_IS_DOUBLE
global using real_t = System.Double;
#else
global using real_t = System.Single;
#endif
```

**Comment by HaloFour** [comment]
I think there's work being done to support [natively-sized floating point](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/NFloat.cs) types in the runtime.  Maybe if there is enough of an interest they can then consider adding an `nfloat` alias to the language.  There's also work to support [generic math](https://devblogs.microsoft.com/dotnet/dotnet-7-generic-math/) on any numeric type, including user-defined, coming in .NET 7.0 and C# 11.

**Comment by tvardero** [comment]
Could see `nfloat` alias to be used in high performant applications, like game engines (where every bit of performance being squeezed out).
It is not much difference from using (System.Runtime.InteropServices.)NFloat, but for sake of perfectionism 😁 (we do have nint for integers)...

Will it be considered for .Net 8?
I don't see posibillity to backport it to prev versions of .Net, since adding a new keyword is a breaking change.

**Comment by The-Futurist** [comment]
> Since C# 9 we have `nint` and `nuint`, but what about `nfloat`? As well as `NMath` library for math operations (sin, cos, round, etc.). I see that Xamarin.Mac SDK [has it already](https://learn.microsoft.com/en-us/dotnet/api/system.nfloat?view=xamarin-mac-sdk-14).
> 
> Current approach is to use this code:
> 
> ```cs
> #if REAL_T_IS_DOUBLE
> global using real_t = System.Double;
> #else
> global using real_t = System.Single;
> #endif
> ```

What is the motive for this alias idea? a 32 bit float is _fundamentally different_ to a 64 bit float, why would we want to treat them as if they were the same?

In the [IEEE 754](https://en.wikipedia.org/wiki/IEEE_754) standard they define these types with names "binary32" and "binary64", I'm just asking because I don't understand where this requirement comes from.

**Reply by HaloFour** [reply]
See: https://github.com/dotnet/runtime/issues/4233

TL;DR: on some platforms, like MacOS/iOS, a lot of APIs are defined as accepting native platform-sized data types.  .NET has always supported that for integers (via `System.IntPtr` and later `nint`/`nuint`) but that doesn't address similar APIs for floating point operations.
